### PR TITLE
#2002 Early removal of a replaced placeholder parameter in `DownstreamUrlCreatorMiddleware`

### LINKED ISSUE #2002: Regression at DownstreamUrlCreatorMiddleware
## Expected Behavior
There is a regression since version 22.0.0 and even in version 23.0.0.
```json
"UpstreamPathTemplate": "/WeatherForecast/{roleId}/groups?something=9874565",
"UpstreamHttpMethod": ["Get"],
"DownstreamPathTemplate": "/account/{{username}}/groups/{{groupName}}/roles?roleId={roleId}",
"DownstreamScheme": "https",
"DownstreamHostAndPorts": [
  { "Host": "localhost", "Port": 7275 }
],
```

The correct behabior is :
```
Upstream: /WeatherForecast/123456/groups?something=9874565
Downstream: /account/{{username}}/groups/{{groupName}}/roles?roleid=123456&something=9874565
```

## Actual Behavior / Motivation for New Feature

But currently all query from Downstream path was removed

## Steps to Reproduce the Problem

Make sure that you have a query from DownstreamPathTemplate and UpstreamPathTemplate
```json
"UpstreamPathTemplate": "/WeatherForecast/{roleId}/groups?something=9874565",
"UpstreamHttpMethod": ["Get"],
"DownstreamPathTemplate": "/account/{{username}}/groups/{{groupName}}/roles?roleId={roleId}",
"DownstreamScheme": "https",
"DownstreamHostAndPorts": [
  { "Host": "localhost", "Port": 7275 }
],
```

## Specifications

  - Version: all version from 22.0.0+

### PR DESCRIPTION
**Pull request by bbenameur**
## Fixes #2002
- #2002 

## Proposed Changes

Remove only added query to placeholders.

### CODE REVIEW

**Comment by raman-m** [comment]
I answered you in #2002, see [link](https://github.com/ThreeMammals/Ocelot/issues/2002#issuecomment-2015468642)!
This PR fixes your rare user case when you wanted just use the same names of placeholder and query param.
So, you wanted to use `userId` placeholder in query string like `?userId={userId}`.
Ocelot has a list of other features in [Placeholders](https://ocelot.readthedocs.io/en/latest/features/routing.html#placeholders) logic. And probably your PR will break them...
Placeholders logic is quite complex and stable now, and **we don't want to change the logic at all!**
Read [the docs](https://ocelot.readthedocs.io/en/latest/features/routing.html#placeholders) for correct usage.

As a hotfix DevOps you need to rollback to version 21.0 as I suggested to you.

If you need something more, then just fork Ocelot repo and develop whatever you want!
Good luck!

**Comment by raman-m** [comment]
Strange... the build [3669](https://circleci.com/gh/ThreeMammals/Ocelot/3669) is 🟢 
That means the unit tests were passed.
So, that means there is no much sense in moving of the code block...

**Comment by bbenameur** [comment]
> Strange... the build [3669](https://circleci.com/gh/ThreeMammals/Ocelot/3669) is 🟢 That means the unit tests were passed. So, that means there is no much sense in moving of the code block...

The build https://app.circleci.com/jobs/github/ThreeMammals/Ocelot/3669 is green because the new code have no regrission and work correctly , but we need some new test that cover the case where param name and value are some like
userId=userId

**Comment by raman-m** [comment]
>  but we need some new test that cover the case where param name and value are some like
userId=userId

🆗 Work more!
Our dev process requires tests for sure. Include unit tests and one acceptance test please.

File: src/Ocelot/DownstreamUrlCreator/Middleware/DownstreamUrlCreatorMiddleware.cs [hunk 1]
```diff
@@ -98,6 +98,15 @@ private static string MergeQueryStringsWithoutDuplicateValues(string queryString
             newQueryString = newQueryString.Replace(QuestionMark, Ampersand);
             var queries = HttpUtility.ParseQueryString(queryString);
             var newQueries = HttpUtility.ParseQueryString(newQueryString);
+
+            // Remove old replaced query parameters
+            foreach (var queryKey in queries.AllKeys)
+            {
+                foreach (var unused in placeholders.Where(placeholder => queryKey == placeholder.Name.Trim(OpeningBrace, ClosingBrace)))
+                {
+                    queries.Remove(queryKey);
```
Comment (RaynaldM) [inline src/Ocelot/DownstreamUrlCreator/Middleware/DownstreamUrlCreatorMiddleware.cs:107]:
```suggestion
 var matchingPlaceholder = placeholders.FirstOrDefault(placeholder => queryKey == placeholder.Name.Trim(OpeningBrace, ClosingBrace));
    if (matchingPlaceholder != null)
    {
        queries.Remove(queryKey);
    }
```

**Review by RaynaldM — APPROVED** [review]

File: src/Ocelot/DownstreamUrlCreator/Middleware/DownstreamUrlCreatorMiddleware.cs (hunk 1 above)
Comment (bbenameur) [reply src/Ocelot/DownstreamUrlCreator/Middleware/DownstreamUrlCreatorMiddleware.cs:107]: @RaynaldM, done

File: src/Ocelot/DownstreamUrlCreator/Middleware/DownstreamUrlCreatorMiddleware.cs [hunk 2]
```diff
@@ -98,6 +98,15 @@ private static string MergeQueryStringsWithoutDuplicateValues(string queryString
...
             var newQueries = HttpUtility.ParseQueryString(newQueryString);
+
+            // Remove old replaced query parameters
+            foreach (var queryKey in queries.AllKeys)
+            {
+                foreach (var unused in placeholders.Where(placeholder => queryKey == placeholder.Name.Trim(OpeningBrace, ClosingBrace)))
+                {
+                    queries.Remove(queryKey);
+                }
+            }
```
Comment (ggnaegi) [inline src/Ocelot/DownstreamUrlCreator/Middleware/DownstreamUrlCreatorMiddleware.cs:109]:
```suggestion
           var placeholderNames = new HashSet<string>(placeholders.Select(p => p.Name.Trim(OpeningBrace, ClosingBrace)));
foreach (var queryKey in queries.AllKeys.Where(placeholderNames.Contains).ToList())
{
    queries.Remove(queryKey);
}
```

Comment (ggnaegi) [reply src/Ocelot/DownstreamUrlCreator/Middleware/DownstreamUrlCreatorMiddleware.cs:109]: We might win some cpu cycles here, up to you

**Review by ggnaegi — COMMENTED** [review]
I have added an alternate method, up to you.

File: src/Ocelot/DownstreamUrlCreator/Middleware/DownstreamUrlCreatorMiddleware.cs (hunk 2 above)
Comment (bbenameur) [reply src/Ocelot/DownstreamUrlCreator/Middleware/DownstreamUrlCreatorMiddleware.cs:109]: Why the .ToList() ?

Comment (ggnaegi) [reply src/Ocelot/DownstreamUrlCreator/Middleware/DownstreamUrlCreatorMiddleware.cs:109]: @bbenameur yes you can remove that, I have tried several strategies and at some point Resharper complained about multiple enumerations.

**Comment by raman-m** [comment]
> @ggnaegi left [a comment](https://github.com/ThreeMammals/Ocelot/pull/2003#pullrequestreview-1975203966)

Do you approve or not?

File: test/Ocelot.AcceptanceTests/RoutingWithQueryStringTests.cs [hunk 3]
```diff
@@ -312,6 +312,44 @@ public void Should_return_response_200_with_query_string_upstream_template_multi
...
+                        DownstreamScheme = "http",
+                        DownstreamHostAndPorts =
+                        [
+                            new()
+                            {
+                                Host = "localhost",
+                                Port = port,
+                            },
+                        ],
+                        UpstreamPathTemplate = "/api/subscriptions/{subscriptionId}/updates?unitId={unitId}",
```
Comment (raman-m) [inline test/Ocelot.AcceptanceTests/RoutingWithQueryStringTests.cs:339]: This is copied and pasted test (user scenario) of old tests.
You should not copy old tests at all!
**You have to show us that your user scenario is fixed by your code improvement!**

File: test/Ocelot.UnitTests/DownstreamUrlCreator/DownstreamUrlCreatorMiddlewareTests.cs [hunk 4]
```diff
@@ -543,6 +543,70 @@ public void should_fix_issue_748(string upstreamTemplate, string downstreamTempl
...
+                .WithDownstreamScheme(Uri.UriSchemeHttp)
+                .Build();
+            var config = new ServiceProviderConfigurationBuilder().Build();
+
+            this.Given(x => x.GivenTheDownStreamRouteIs(
+                    new DownstreamRouteHolder(
+                        new List<PlaceholderNameAndValue>
+                        {
+                            new("{userId}", "webley"),
+                            new("{personId}", "12345"),
```
Comment (raman-m) [inline test/Ocelot.UnitTests/DownstreamUrlCreator/DownstreamUrlCreatorMiddlewareTests.cs:564]: ### Again! :grimacing: 
This is copied user scenario from old tests!
**Where is testing of your user scenario :question:**

File: test/Ocelot.UnitTests/DownstreamUrlCreator/DownstreamUrlCreatorMiddlewareTests.cs [hunk 5]
```diff
@@ -543,6 +543,70 @@ public void should_fix_issue_748(string upstreamTemplate, string downstreamTempl
...
+                        new RouteBuilder().WithDownstreamRoute(downstreamRoute)
+                            .WithUpstreamHttpMethod(methods)
+                            .Build())))
+                .And(x => x.GivenTheDownstreamRequestUriIs($"http://localhost:5000/users?userId=webley"))
+                .And(x => GivenTheServiceProviderConfigIs(config))
+                .And(x => x.GivenTheUrlReplacerWillReturn("/persons?groupId=6789&personId=12345&userId=webley"))
+                .When(x => x.WhenICallTheMiddleware())
+                .Then(x => x.ThenTheDownstreamRequestUriIs($"http://localhost:5000/persons?groupId=6789&personId=12345&userId=webley"))
+                .And(x => ThenTheQueryStringIs("?groupId=6789&personId=12345&userId=webley"))
+                .BDDfy();
```
Comment (raman-m) [inline test/Ocelot.UnitTests/DownstreamUrlCreator/DownstreamUrlCreatorMiddlewareTests.cs:607]: We don't use BDDfy for new unit tests anymore! See `UnitTest` class constructor!
BDDfy will be removed from all unit tests soon... This is absolutely useless framework for unit tests. xUnit works well...

File: test/Ocelot.UnitTests/DownstreamUrlCreator/DownstreamUrlCreatorMiddlewareTests.cs [hunk 6]
```diff
@@ -543,6 +543,70 @@ public void should_fix_issue_748(string upstreamTemplate, string downstreamTempl
...
+        }
+
+        [Fact]
+        public void Should_map_all_given_traffic_query_with_placeholder()
+        {
+            var methods = new List<string> { "Post", "Get" };
+            var downstreamRoute = new DownstreamRouteBuilder()
+                .WithUpstreamPathTemplate(new UpstreamPathTemplateBuilder()
+                    .WithOriginalValue("/users?userId={userId}&personId={personId}&groupId={groupId}").Build())
+                .WithDownstreamPathTemplate("/persons?personId={personId}&userId={userId}")
```
Comment (raman-m) [inline test/Ocelot.UnitTests/DownstreamUrlCreator/DownstreamUrlCreatorMiddlewareTests.cs:585]: I wonder if this scenario and test is green now...
We have a conflict with documentation now.
Published feature will not be removed from docs.
You may not like it, but it works well, and fixes a lot of old bugs.

**Review by raman-m — CHANGES_REQUESTED** [review]
The code changes can break functionality of [Query String Placeholders feature](https://github.com/ThreeMammals/Ocelot/blob/develop/docs/features/routing.rst#query-string-placeholders) of placeholder replacements in query string. I need to understand why old tests green and final decision will be made after realizing that published functionality of [Restrictions on use](https://github.com/ThreeMammals/Ocelot/blob/develop/docs/features/routing.rst#restrictions-on-use) still work well.
Now I realized that the name of this docs section confuses community but actually this is the feature which will not be removed from the docs, but I will give the new name to the section.

File: src/Ocelot/DownstreamUrlCreator/Middleware/DownstreamUrlCreatorMiddleware.cs (hunk 2 above)
Comment (raman-m) [reply src/Ocelot/DownstreamUrlCreator/Middleware/DownstreamUrlCreatorMiddleware.cs]: Fixed in commit https://github.com/ThreeMammals/Ocelot/pull/2003/commits/c43030ca4dcee8be1d3be665be5bb4b535b2db75 ✔️

**Review by raman-m — APPROVED** [review]
### Ready for delivery ✅ 
- Code review ✔️✔️
- Team approvals ✔️✔️
- Unit testing ✔️ &rarr; `DownstreamUrlCreatorMiddlewareTests` &rarr; [link1](https://github.com/ThreeMammals/Ocelot/pull/2003/files#diff-bdac7f678a94c0613391e5944148744ce139f6a6b95f24883c4d53198e05c4eb)
- Acceptance testing ✔️ &rarr; `RoutingWithQueryStringTests` &rarr; [link2](https://github.com/ThreeMammals/Ocelot/pull/2003/files#diff-8761b9b50aa53952b5f8ee3a400af51cd2b6755fc4546b1cce35d8dad5d9023d)
- Updated docs ✔️ &rarr; `routing.rst` &rarr; [link3](https://github.com/ThreeMammals/Ocelot/pull/2003/files#diff-f8cdcb9e36448707f046f3a7ee0f861135cef70cd2fe0153a82a0873bf68c45e)

**Pull request merged**